src/sync/spsc_bufferqueue.rs

```diff
@@ -0,0 +1,400 @@
+//! SPSC ringbuffer
+
+use std::sync::atomic::Ordering::{Acquire, Release, Relaxed, SeqCst};
+use std::sync::atomic::{AtomicUsize, AtomicBool, fence};
+use std::sync::Arc;
+use std::ptr;
+use std::mem;
+use std::cmp;
+use std::marker::PhantomData;
+use mem::CachePadded;
+
+#[inline(always)]
+unsafe fn deallocate<T>(ptr: *mut T, old_size: usize) {
+    Vec::from_raw_parts(ptr, 0, old_size);
+}
+
+#[inline(always)]
+unsafe fn allocate<T>(size: usize) -> *mut T {
+    let mut buf = Vec::with_capacity(size);
+    let ptr = buf.as_mut_ptr();
+    mem::forget(buf);
+    ptr
+}
+
+/// A single-producer, single consumer bounded wait-free ringbuffer queue
+///
+/// All operations on the buffer queue are wait-free,
+/// provided move/copy operations are waitfree.
+/// This queue does not allocate after constructions
+#[repr(C)]
+pub struct SpscBufferQueue<T: Send> {
+    // This is a pointer instead of a vector
+    // so that vector doesn't call constructors
+    data_block: *mut T,
+    size: usize,
+    _marker: PhantomData<T>,
+
+    // These dummies result in a tremendous performance improvement, ~300%+
+    _dummy_1: CachePadded<u64>,
+    // data for the consumer
+    head: AtomicUsize,
+    tail_cache: AtomicUsize,
+    prod_alive:AtomicBool, //seems weird, but consumer will read this
+
+    _dummy_2: CachePadded<u64>,
+    // data for the producer
+    tail: AtomicUsize,
+    head_cache: AtomicUsize,
+    cons_alive: AtomicBool, //seems weird, but producer will read this
+}
+
+unsafe impl<T: Send> Send for SpscBufferQueue<T> {}
+
+impl<T: Send> SpscBufferQueue<T> {
+    pub fn new(size: usize) -> (BufferProducer<T>, BufferConsumer<T>) {
+        let q = SpscBufferQueue {
+            data_block: unsafe{ allocate(size) },
+            size: cmp::min(size, (isize::max_value() - 1) as usize) + 1,
+            _marker: PhantomData,
+
+            _dummy_1: CachePadded::zeroed(),
+            head: AtomicUsize::new(0),
+            tail_cache: AtomicUsize::new(0),
+            prod_alive: AtomicBool::new(true),
+
+            _dummy_2: CachePadded::zeroed(),
+            tail: AtomicUsize::new(0),
+            head_cache: AtomicUsize::new(0),
+            cons_alive: AtomicBool::new(true),
+        };
+        let qarc = Arc::new(q);
+        let rtuple = (BufferProducer::new(qarc.clone()),
+                      BufferConsumer::new(qarc));
+        fence(SeqCst);
```

Comment:
It originally was so that users wouldn't have to think about ordering when sharing the queue across threads, but I realized now thinking about it either an external mechanism will be used or someone would have to enforce ordering on both the construction and queue load, not just construction. It's an edge case anyways and is much harder to trigger in Rust anyways without realizing it.